core/src/main/java/hivemall/tools/map/MapRouletteUDF.java

```diff
@@ -49,164 +39,171 @@
+            if (HiveUtils.isIntegerOI(argOI1) == false) {
                 throw new UDFArgumentException(
-                    "Expected a number but get: " + valueOI.getTypeName());
+                    "The second argument of map_roulette must be integer type: "
+                            + argOI1.getTypeName());
+            }
+            if (ObjectInspectorUtils.isConstantObjectInspector(argOI1)) {
+                long seed = HiveUtils.getAsConstLong(argOI1);
+                this._rand = new Random(seed); // fixed seed
+            } else {
+                this.seedOI = HiveUtils.asLongCompatibleOI(argOI1);
+            }
+        } else {
+            this._rand = new Random(); // random seed
         }
-        return keyOI;
+
+        return mapOI.getMapKeyObjectInspector();
     }
 
     @Override
     public Object evaluate(DeferredObject[] arguments) throws HiveException {
-        Map<Object, Double> input = processObjectDoubleMap(arguments[0]);
+        Random rand = _rand;
+        if (rand == null) {
+            Object arg1 = arguments[1].get();
+            if (arg1 == null) {
+                rand = new Random();
+            } else {
+                long seed = HiveUtils.getLong(arg1, seedOI);
+                rand = new Random(seed);
+            }
+        }
+
+        Map<Object, Double> input = getObjectDoubleMap(arguments[0], mapOI, valueOI);
         if (input == null) {
             return null;
         }
-        return algorithm(input);
-    }
 
-    @Override
-    public String getDisplayString(String[] children) {
-        return "map_roulette(" + StringUtils.join(children, ',') + ")";
+        return rouletteWheelSelection(input, rand);
     }
 
-    /**
-     * Process the data passed by user.
-     * 
-     * @param argument data passed by user
-     * @return If all the value is ,
-     * @throws HiveException If get the wrong weight value like {key = "Wang", value = "Zhang"},
-     *         "Zhang" isn't a number ,this Method will throw exception when
-     *         convertPrimitiveToDouble("Zhang", valueOD)
-     */
-    private Map<Object, Double> processObjectDoubleMap(DeferredObject argument)
-            throws HiveException {
-        // get
-        Map<?, ?> m = mapOI.getMap(argument.get());
+    @Nullable
+    private static Map<Object, Double> getObjectDoubleMap(@Nonnull final DeferredObject argument,
+            @Nonnull final MapObjectInspector mapOI,
+            @Nonnull final PrimitiveObjectInspector valueOI) throws HiveException {
+        final Map<?, ?> m = mapOI.getMap(argument.get());
         if (m == null) {
             return null;
         }
-        if (m.size() == 0) {
+        final int size = m.size();
+        if (size == 0) {
             return null;
         }
-        // convert
-        Map<Object, Double> input = new HashMap<>();
-        Double avg = 0.0;
+
+        final Map<Object, Double> result = new HashMap<>(size);
```

Comment:
`Map<Object, Double> result` will be changed, why use final here?

Reply:
I cannot catch you. 

This instance is not changed and returned on `getObjectDoubleMap()`. The annotation means that `result` is not replaced.

`final` annotation for a map instance does not mean that `result` can not be updated.

Reply:
Got it, sorry for my poor English……